keycloak/keycloak-operator
# KEYCLOAK-14774 Add AuthenticationFlows and AuthenticatorConfig to KeycloakRealm CRD

## KEYCLOAK-14774


## Additional Information
Why: Allow custom authentication flows

## Verification Steps
1. Create a Realm with custom authentication flows
2. Verify in the admin UI that they have been imported

```yaml
spec:
  instanceSelector:
    matchLabels:
      app: sso
  realm:
    displayName: Realm
    enabled: true
    id: oidc
    realm: oidc
    authenticatorConfig:
      - alias: "oidc"
        config:
          defaultProvider: "oidc"
    authenticationFlows:
    - alias: "Auto Link"
      providerId: "basic-flow"
      topLevel: true
      builtIn: false
      authenticationExecutions:
      - authenticator: "idp-create-user-if-unique"
        requirement: "ALTERNATIVE"
        priority: 0
        userSetupAllowed: false
        autheticatorFlow: false
      - authenticator: "idp-auto-link"
        requirement: "ALTERNATIVE"
        priority: 1
        userSetupAllowed: false
        autheticatorFlow: false
    - alias: "browser"
      providerId: "basic-flow"
      topLevel: true
      builtIn: false
      authenticationExecutions:
      - authenticator: "identity-provider-redirector"
        authenticatorConfig: oidc
        requirement: "ALTERNATIVE"
        priority: 0
        userSetupAllowed: false
        autheticatorFlow: false
    - alias: "registration"
      providerId: "basic-flow"
      topLevel: true
      builtIn: false
      authenticationExecutions: []
    - alias: "direct grant"
      providerId: "basic-flow"
      topLevel: true
      builtIn: false
      authenticationExecutions: []
```
## Checklist:

## Additional Notes 
Two server-side issues have been created:
* "registration" and "direct grant" flow should be auto-created if missing: https://issues.redhat.com/browse/KEYCLOAK-14779
* Do not NPE on missing authenticatorConfig: https://issues.redhat.com/browse/KEYCLOAK-14780
